core/foundation/inc/ROOT/RLogger.hxx

```diff
@@ -82,90 +212,101 @@ public:
+   /// Get the number of warnings that the RLogDiagCount has emitted since construction of *this.
+   long long GetAccumulatedWarnings() const { return fCounter->GetNumWarnings() - fInitialWarnings; }
 
-   /// Whether the RLogManager has emitted an error since construction time of *this.
-   bool HasErrorOccurred() const { return GetAccumulatedErrors(); }
+   /// Get the number of errors that the RLogDiagCount has emitted since construction of *this.
+   long long GetAccumulatedErrors() const { return fCounter->GetNumErrors() - fInitialErrors; }
 
-   /// Whether the RLogManager has emitted an error or a warning since construction time of *this.
-   bool HasErrorOrWarningOccurred() const { return HasWarningOccurred() || HasErrorOccurred(); }
-};
+   /// Get the number of errors that the RLogDiagCount has emitted since construction of *this.
+   long long GetAccumulatedFatalErrors() const { return fCounter->GetNumFatalErrors() - fInitialFatalErrors; }
 
-/**
- A diagnostic, emitted by the RLogManager upon destruction of the RLogEntry.
- One can construct a RLogEntry through the utility preprocessor macros R__ERROR_HERE, R__WARNING_HERE etc
- like this:
-     R__INFO_HERE("CodeGroupForInstanceLibrary") << "All we know is " << 42;
- This will automatically capture the current class and function name, the file and line number.
- */
+   /// Whether the RLogDiagCount has emitted a warnings since construction time of *this.
+   bool HasWarningOccurred() const { return GetAccumulatedWarnings(); }
 
-class RLogEntry: public std::ostringstream {
-public:
-   std::string fGroup;
-   std::string fFile;
-   std::string fFuncName;
-   int fLine = 0;
-   ELogLevel fLevel;
+   /// Whether the RLogDiagCount has emitted an error (fatal or not) since construction time of *this.
+   bool HasErrorOccurred() const { return GetAccumulatedErrors() + GetAccumulatedFatalErrors(); }
 
-public:
-   RLogEntry() = default;
-   RLogEntry(ELogLevel level, std::string_view group): fGroup(group), fLevel(level) {}
-   RLogEntry(ELogLevel level, std::string_view group, std::string_view filename, int line, std::string_view funcname)
-      : fGroup(group), fFile(filename), fFuncName(funcname), fLine(line), fLevel(level)
-   {}
+   /// Whether the RLogDiagCount has emitted an error or a warning since construction time of *this.
+   bool HasErrorOrWarningOccurred() const { return HasWarningOccurred() || HasErrorOccurred(); }
+};
 
-   RLogEntry &SetFile(const std::string &file)
+namespace Internal {
+   /**
+    Utility class that prevents stream operators from being called if verbosity is too low, i.e.:
+    ````
+    RLogScopedVerbosity silence(RLogLevel::kFatal);
+    R__DEBUG_HERE({}) << WillNotBeCalled();
+    ```
+    */
+   struct TagForDefaultVerbosity {};
+   inline bool ShouldCreateLogEntry(ELogLevel level, TagForDefaultVerbosity)
    {
-      fFile = file;
-      return *this;
+      return RLogManager::Get().GetVerbosity() >= level || level < ELogLevel::kInfo;
```

Comment:
How do you suppress Warning message?

Reply:
`RLogScopedVerbosity silence(ELogLevel::kError)`, unless the channel has warnings turned on. (That's just the answer to your question - we can discuss the interplay of global vs channel verbosity.)

Reply:
Can you rephrase? I don't understand the second sentence. I'm okay with the first one ;-)

Reply:
Users might be surprised that `RLogManager::Get().SetVerbosity(ELogLevel::kError)` leads to warning message being still printed while `RLogScopedVerbosity silence(ELogLevel::kError)` put in the `main function` removes them.  (Unless I misunderstood something)

Reply:
`RLogScopedVerbosity silence(ELogLevel::kError)` is just two calls to `RLogManager::Get().SetVerbosity(ELogLevel::kError)`, taking care of resetting to whatever the verbosity was before. I.e.
```
void f() {
  RLogScopedVerbosity silence(ELogLevel::kError);
  Sometihng();
}
```
and
```
void f() {
  auto prev = RLogManager::Get().SetVerbosity(ELogLevel::kError);
  Something();
  RLogManager::Get().SetVerbosity(prev);
}
```
are strictly identical.

Whether a channel's log entries are emitted or not is an "OR" of the global and the channel-specific verbosity: whatever is more verbose wins. Example:

```
RLogChannel ch("Test");
R__WARNING_HERE(ch) << "Emitted";
{
  RLogScopedVerbosity(kError);
  R__WARNING_HERE(ch) << "Suppressed";
}
{
  RLogScopedVerbosity(ch, kError);
  R__WARNING_HERE(ch) << "Emitted"; // global wins - surprisingly?
}
{
  RLogScopedVerbosity(kError);
  RLogScopedVerbosity(ch, kWarning);
  R__WARNING_HERE(ch) << "Visible"; // this time, channel setting wins
}
```

Rationale: allow people to get a silent ROOT - the global verbosity set to `kFatal` should really only show fatal errors unless verbosity is increased for specific channels in which case these channels' logs will be emitted.

One could also argue that a channel's specific setting wins over the global one, and introduce an "unset" channel verbosity, in which case the global verbosity it taken:

```
RLogChannel ch("Test");
R__WARNING_HERE(ch) << "Emitted";
{
  RLogScopedVerbosity(kError);
  R__WARNING_HERE(ch) << "Suppressed";
}
{
  RLogScopedVerbosity(ch, kError);
  R__WARNING_HERE(ch) << "Suppressed"; // channel setting always wins
}
{
  RLogScopedVerbosity(kError);
  RLogScopedVerbosity(ch, kWarning);
  R__WARNING_HERE(ch) << "Visible";
}
```

My feeling was that it might be easier to tell users: "can you run once with `RLogManager::Get().SetVerbosity(ELogLevel::kDebug);`? What's the output?" instead of having to turn on a couple of channels. But I guess that can be done in either scenario. So - shall I change to the second approach because it's less surprising and still allows to do the same things (@pcanal @eguiraud @jblomer mostly, I guess)?

Reply:
I don't have a lot of experience with logging but "channel-specific wins over global" seems strictly more useful than "more verbose wins" because it _also_ allows running e.g. with "everything on _except_ IO".

However like Philippe in the first comment I'm also confused by the `|| level < ELogLevel::kInfo;`: why do we unconditionally want to create the `RLogEntry` if it's a warning or an error (even if we later won't print them out)? The line probably deserves a comment.